{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Neo4J v1 - Trellis : Coherence check results (QC)\n",
    "\n",
    "Coherence check includes results generated by FastQC, Samtools Flagstat, and RTG Vcfstats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip3 install seaborn\n",
    "!pip3 install pandas-gbq\n",
    "!pip3 install google-cloud-storage\n",
    "!pip3 install datalab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "\n",
    "import pandas as pd\n",
    "import pandas_gbq\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from google.cloud import storage\n",
    "import yaml\n",
    "import google.datalab.bigquery as bq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Neo4J DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "bucket_info=''\n",
    "credential_info=''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 1 : Read DB and Account Information in Google Storage (YAML)\n",
    "\n",
    "# create storage client\n",
    "storage_client = storage.Client()\n",
    "# get bucket with name\n",
    "bucket = storage_client.get_bucket(bucket_inf)\n",
    "# get bucket data as blob\n",
    "blob = bucket.get_blob(credential_info)\n",
    "# convert to string\n",
    "yaml_data = blob.download_as_string()\n",
    "\n",
    "account = yaml.load(yaml_data, Loader=yaml.FullLoader)\n",
    "\n",
    "## Main Account\n",
    "#graph = Graph(account['NEO4J_SCHEME']+'://'+account['NEO4J_HOST']+\":\"+str(account['NEO4J_PORT']), auth=(account['NEO4J_USER'],account['NEO4J_PASSPHRASE']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FastQC Results (n=285)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bigquery table (FastQC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_fastqc_table=account['GOOGLE_CLOUD_PROJECT']+'.'+account['BIGQUERY_DATASET']+'.fastqc'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per base quality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inserted aggregation function (Average) for per base quality of each sample based on the FastQC results of 25 chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_base_qual = bq.Query('SELECT index, value_tenth, ' +\n",
    "                         'COUNT(*) AS observations ' +\n",
    "                         'FROM (SELECT sample, index, ' +\n",
    "                         'ROUND((CAST(value AS FLOAT64)), 1) AS value_tenth, ' +\n",
    "                         'FROM `' + mvp_fastqc_table + '` ' +\n",
    "                         'WHERE dimension = \"per_base_qual\") ' +\n",
    "                         'GROUP BY index, value_tenth').execute().result().to_dataframe()\n",
    "#per_base_qual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_base_qual.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Divided start point and end point of base quality index  <br />\n",
    "Changed data type from string to num  <br />\n",
    "Added log_observation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_end = pd.DataFrame(per_base_qual['index'].str.split('-',1).tolist(), columns=['start', 'end'])\n",
    "per_base_qual = per_base_qual.join(start_end)\n",
    "\n",
    "per_base_qual[['value_tenth']] = per_base_qual[['value_tenth']].apply(pd.to_numeric)\n",
    "per_base_qual[['start']] = per_base_qual[['start']].apply(pd.to_numeric)\n",
    "per_base_qual[['observations']] = per_base_qual[['observations']].apply(pd.to_numeric)\n",
    "\n",
    "per_base_qual[['log_observations']] = np.log(per_base_qual[['observations']])\n",
    "#per_base_qual.head()\n",
    "#per_base_qual.to_csv('per_base_qual.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of per base quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "scat = per_base_qual.plot(kind='scatter', x='start', y='value_tenth', c='log_observations', cmap='plasma', ax=ax)\n",
    "ax.set_xlim(0,170)\n",
    "ax.set_ylim(0,35)\n",
    "ax.set_title(\"Distribution of Per-base Sequence Quality across Genomes\", size=15)\n",
    "ax.set_xlabel(\"Base position\", size=15)\n",
    "ax.set_ylabel(\"Quality score\", size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GC Content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gc_content value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"gc_content\"\n",
    "gc_content = bq.Query('SELECT sample, CAST(value as INT64) AS GC_value ' + \n",
    "                      'FROM `' + mvp_fastqc_table + '` ' +\n",
    "                      'WHERE dimension = \"' + dimension + '\"' ).execute().result().to_dataframe()\n",
    "display(gc_content.shape[0])\n",
    "#gc_content.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gc_outlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean=gc_content.GC_value.mean()\n",
    "gc_content['gc_outlier'] = gc_content.GC_value.apply(lambda x: False if x in [40,41] else True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_gc=gc_content.rename(columns={\"GC_value\":\"gc\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution of GC content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "gc_content.hist(alpha=0.9, ax=ax, bins=5)\n",
    "ax.set_xlim(39,42)\n",
    "#ax.set_ylim(0,250)\n",
    "ax.set_title(\"Distribution of Reads GC Content by Genome\",size=15)\n",
    "ax.set_xlabel(\"GC Content (%)\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Average sequence quality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sum seq_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"seq_quality\"\n",
    "avg_seq_qual = bq.Query('SELECT sample, ' +\n",
    "                        'SUM(CAST(index as INT64) * CAST(value AS FLOAT64))/SUM(CAST(value AS FLOAT64)) ' + \n",
    "                        'FROM `' + mvp_fastqc_table + '` ' + \n",
    "                        'WHERE dimension = \"' + dimension + '\" ' +\n",
    "                        'GROUP BY sample').execute().result().to_dataframe()\n",
    "\n",
    "#print(avg_seq_qual.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#qc_bq=gc_content.rename(columns={\"GC_value\":\"gc\"})\n",
    "std=avg_seq_qual.f0_.std()\n",
    "mean=avg_seq_qual.f0_.mean()\n",
    "#avg_seq_qual['avg_seq_qual_outlier'] = avg_seq_qual.f0_.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "avg_seq_qual['avg_seq_qual_outlier'] = avg_seq_qual.f0_.apply(lambda x: False if x > 28 else True)\n",
    "bq_avg_seq_qual=avg_seq_qual.rename(columns={\"f0_\":\"avg_seq_qual\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### statistics of average sequence quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Mean of sequence quality: {0:0.2f}'.format(np.mean(avg_seq_qual.f0_)))\n",
    "print('Median of sequence quality: {0:0.2f}'.format(np.median(avg_seq_qual.f0_)))\n",
    "print('STD of sequence quality: {0:0.2f}'.format(np.std(avg_seq_qual.f0_)))\n",
    "avg_seq_qual.head()\n",
    "avg_seq_qual.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "avg_seq_qual['avg_seq_qual_outlier'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution of average sequence quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "avg_seq_qual.f0_.hist(alpha=0.9, ax=ax)\n",
    "ax.set_xlim(28,30)\n",
    "ax.set_title(\"Distribution of Average Read Qualities\",size=15)\n",
    "ax.set_xlabel(\"Average Read Quality\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"seq_len\"\n",
    "seq_len = bq.Query('SELECT sample, dimension, ' +\n",
    "                   'CAST(value AS INT64) as value '\n",
    "                   'FROM `' + mvp_fastqc_table + '` ' + \n",
    "                   'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "#seq_len.head()\n",
    "seq_len.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_len_counts = seq_len.groupby('value')['sample'].nunique()\n",
    "print (seq_len_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_len['seq_len_outlier'] = seq_len.value.apply(lambda x: False if x == 150 else True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of sequence length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "bar = seq_len_counts.plot.bar(alpha=0.9)\n",
    "ax.set_title(\"Distribution of Read Lengths\")\n",
    "ax.set_xlabel(\"Read Length\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_seq_len=seq_len.loc[:,['sample','value','seq_len_outlier']].rename(columns={\"value\":\"seq_len\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VCFStats Results (n=288)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bigquery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_vcfstats_table = account['GOOGLE_CLOUD_PROJECT']+'.'+account['BIGQUERY_DATASET']+'.vcfstats'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNPs per genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of samples <br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = bq.Query('SELECT sample, dimension, ' +\n",
    "                'CAST(value AS INT64) as value ' +\n",
    "                'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                'WHERE dimension = \"snps\" ').execute().result().to_dataframe()\n",
    "#print(snps.head())\n",
    "print('The number of samples: '+str(snps.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics : mean, std, outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "std=snps.value.std()\n",
    "mean=snps.value.mean()\n",
    "print('mean: {0},  std: {1}'.format(round(mean,2), round(std,2)))\n",
    "\n",
    "# if the measurement is 3 times STD from the mean, then flag it\n",
    "snps['Outlier'] = snps.value.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "print('The number of outliers: {0}'.format(snps[snps.Outlier].shape[0]))\n",
    "print(snps[snps.Outlier])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = snps.value.hist(alpha=0.9, ax=ax, bins=10)\n",
    "ax.set_xlim(3000000,5500000)\n",
    "ax.set_title(\"Distribution of SNPs per Genome\",size=15)\n",
    "ax.set_xlabel(\"SNP Count\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()\n",
    "#plt.savefig('Distribution of SNPs per Genome.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_snps=snps.loc[:,['sample','value','Outlier']]. \\\n",
    "    rename(columns={\"value\":\"snps\",\"Outlier\":\"snps_outlier\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check number of variants in each category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension_sums = bq.Query('SELECT dimension, SUM(CAST(value AS INT64)) AS sum ' +\n",
    "                'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                'WHERE dimension IN (\"failed_filters\", \"passed_filters\") ' +\n",
    "                'GROUP BY dimension').execute().result().to_dataframe()\n",
    "dimension_sums.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indels per genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "dimension1 = \"indels\"\n",
    "dimension2 = \"insertions\"\n",
    "dimension3 = \"deletions\"\n",
    "\n",
    "indels = bq.Query('SELECT sample, ' +\n",
    "                  'SUM(CAST(value AS INT64)) as value ' +\n",
    "                  'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                  'WHERE (dimension = \"' + dimension1 + '\") OR (dimension = \"' + dimension2 + '\") OR (dimension = \"' + dimension3 + '\")' +\n",
    "                  'GROUP BY sample').execute().result().to_dataframe()\n",
    "#indels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots()\n",
    "#hist = indels.hist(alpha=0.5, ax=ax)\n",
    "#ax.set_title(\"Distribution of INDELs per Genome\")\n",
    "#ax.set_xlabel(\"INDEL Count\")\n",
    "#ax.set_ylabel(\"Frequency\")\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "#box = indels.plot.box()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "#plt.figure()\n",
    "#plt.boxplot(indels['value'], 0, 'gD')\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cumulative insertions, deletions, and indels\n",
    "RTG Tools vcfstats outputs data on three categories of indels: insertions, deletions, and indels. From this google group post (https://groups.google.com/a/realtimegenomics.com/forum/#!searchin/rtg-users/vcfstats/rtg-users/-eFsSbWF6ks/1HrnevHTAgAJ):\n",
    "\n",
    ">For the insertions/deletions/indels the table is based on the delta in length rather than total length (which really matters for the indels):\n",
    "\n",
    ">Insertions (pure addition of bases)\n",
    ">A -> AT (length 1 insertion)\n",
    ">ATT -> ATTTT (length 2 insertion) \n",
    "\n",
    ">Deletions (pure deletion of bases)\n",
    ">AT -> A (length 1 deletion)\n",
    ">ATTTT -> ATT (length 2 deletion)\n",
    "\n",
    ">Indels (length changing but not pure)\n",
    ">ATT -> CTTT (length 1 indel)\n",
    ">CTTT -> ATT (length 1 indel)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indels = bq.Query('SELECT sample, SUM(DISTINCT CAST(value AS INT64)) AS cum_indels ' +\n",
    "                  'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                  'WHERE dimension = \"insertions\" ' +\n",
    "                  'OR dimension = \"deletions\" ' +\n",
    "                  'OR dimension = \"indels\" ' +\n",
    "                  'GROUP BY sample').execute().result().to_dataframe()\n",
    "#indels.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics : mean, std, outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "std=indels.cum_indels.std()\n",
    "mean=indels.cum_indels.mean()\n",
    "print('mean: {0},  std: {1}'.format(round(mean,2), round(std,2)))\n",
    "\n",
    "# if the measurement is 3 times STD from the mean, then flag it\n",
    "indels['Outlier'] = indels.cum_indels.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "print('The number of outliers: {0}'.format(indels[indels.Outlier].shape[0]))\n",
    "print(indels[indels.Outlier])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of indels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = indels.cum_indels.hist(alpha=0.9, ax=ax, bins = 20)\n",
    "ax.set_xlim(700000,1200000)\n",
    "ax.set_title(\"Distribution of Cumulative insertions, deletions, and indels\",size=15)\n",
    "ax.set_xlabel(\"Cumulative insertions, deletions, and indels Count\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_indels=indels.rename(columns={\"cum_indels\":\"indels\",\"Outlier\":\"indels_outlier\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Frequency of Ti/Tv ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"ti_tv_ratio\"\n",
    "ti_tv = bq.Query('SELECT sample, dimension, ' + \n",
    "                 'CAST(value AS FLOAT64) as value ' + \n",
    "                 'FROM `' + mvp_vcfstats_table + '` ' + \n",
    "                 'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Outlier table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean=ti_tv.value.mean()\n",
    "display(mean)\n",
    "ti_tv['ti_tv_outlier'] = ti_tv.value.apply(lambda x: False if abs(mean-x)<3*std else True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of Ti/Tv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = ti_tv.value.hist(alpha=0.9, ax=ax, bins=8)\n",
    "ax.set_xlim(1.9,2.0)\n",
    "ax.set_title(\"Distribution of Ti/Tv ratios\",size=15)\n",
    "ax.set_xlabel(\"Ti/Tv Ratio\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_ti_tv=ti_tv.loc[:,['sample','value','ti_tv_outlier']]. \\\n",
    "    rename(columns={\"value\":\"ti_tv\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "codeCollapsed": true
   },
   "outputs": [],
   "source": [
    "#plt.figure()\n",
    "#plt.boxplot(ti_tv['value'], 0, 'gD')\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Frequency of SNP het/hom ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_het_hom_ratio = bq.Query('SELECT sample, dimension, ' + \n",
    "                             'CAST(value AS FLOAT64) as value ' + \n",
    "                             'FROM `' + mvp_vcfstats_table + '` ' + \n",
    "                             'WHERE dimension = \"snp_het_hom_ratio\"').execute().result().to_dataframe()\n",
    "#snp_het_hom_ratio.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics : mean, std, outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "std=snp_het_hom_ratio.value.std()\n",
    "mean=snp_het_hom_ratio.value.mean()\n",
    "print('mean: {0},  std: {1}'.format(round(mean,2), round(std,2)))\n",
    "\n",
    "# if the measurement is 3 times STD from the mean, then flag it\n",
    "snp_het_hom_ratio['Outlier'] = snp_het_hom_ratio.value.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "print('The number of outliers: {0}'.format(snp_het_hom_ratio[snp_het_hom_ratio.Outlier].shape[0]))\n",
    "print(snp_het_hom_ratio[snp_het_hom_ratio.Outlier])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of SNP het/hom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = snp_het_hom_ratio.value.hist(alpha=0.9, ax=ax, bins=20)\n",
    "ax.set_title(\"Distribution of SNP het/hom ratios\",size=15)\n",
    "ax.set_xlabel(\"Het/Hom Ratio\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_het_hom=snp_het_hom_ratio.loc[:,['sample','value','Outlier']]. \\\n",
    "    rename(columns={\"value\":\"het_hom\",\"Outlier\":\"het_hom_outlier\"}).set_index('sample')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Samtools Flagstat Results (n=286)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_flagstat_table = account['GOOGLE_CLOUD_PROJECT']+'.'+account['BIGQUERY_DATASET']+'.flagstat'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count and Percent of reads mapped to reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"qc_passed_reads_count\"\n",
    "qc_passed_reads_count = bq.Query('SELECT sample, ' + \n",
    "                             'CAST(value AS FLOAT64) as total_count, ' +\n",
    "                             'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                             'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "print('The number of samples: '+str(qc_passed_reads_count.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"mapped_reads_count\"\n",
    "mapped_reads_count = bq.Query('SELECT sample, ' + \n",
    "                             'CAST(value AS FLOAT64) as mapped_count ' + \n",
    "                             'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                             'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_reads_perc=qc_passed_reads_count.set_index(['sample']).join(mapped_reads_count.set_index(['sample']))\n",
    "mapped_reads_perc['mapped_perc']=100*(mapped_reads_perc['mapped_count']/mapped_reads_perc['total_count'])\n",
    "mapped_reads_perc['total_count']=mapped_reads_perc.total_count.astype(int)\n",
    "mapped_reads_perc['mapped_count']=mapped_reads_perc.mapped_count.astype(int)\n",
    "mapped_reads_perc['mapped_perc']=mapped_reads_perc.mapped_perc.astype(float).round(2)\n",
    "\n",
    "#print('Mean of mapped_reads: {0:0.2f}'.format(np.mean(mapped_reads_perc.mapped_perc)))\n",
    "#print('Median of mapped_reads: {0:0.2f}'.format(np.median(mapped_reads_perc.mapped_perc)))\n",
    "#print('STD of mapped_reads: {0:0.2f}'.format(np.std(mapped_reads_perc.mapped_perc)))\n",
    "#mapped_reads_perc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics : Mean, STD, the number of samples with less than 99% mapped percent "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "std=mapped_reads_perc.mapped_perc.std()\n",
    "mean=mapped_reads_perc.mapped_perc.mean()\n",
    "print('mean: {0},  std: {1}'.format(round(mean,2), round(std,2)))\n",
    "\n",
    "# if the measurement is 3 times STD from the mean, then flag it\n",
    "#mapped_reads_perc['Outlier'] = mapped_reads_perc.mapped_perc.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "mapped_reads_perc['Outlier'] = mapped_reads_perc.mapped_perc.apply(lambda x: False if x > 98 else True)\n",
    "\n",
    "print('The number of samples in < 98% : {0}'.format(mapped_reads_perc[mapped_reads_perc.Outlier].shape[0]))\n",
    "#print('The number of outlier samples : {0}'.format(mapped_reads_perc[mapped_reads_perc.Outlier].shape[0]))\n",
    "print(mapped_reads_perc[mapped_reads_perc.Outlier])\n",
    "\n",
    "mapped_count_outlier = mapped_reads_perc[['mapped_perc','Outlier']]\n",
    "mapped_count_outlier.sort_values(by=['mapped_perc'],inplace=True)\n",
    "#print(mapped_count_outlier[mapped_count_outlier.Outlier])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of Mapped Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = mapped_reads_perc.mapped_count.hist(alpha=0.9, ax=ax)\n",
    "ax.set_title(\"Distribution of Mapped Reads (Count)\")\n",
    "ax.set_xlabel(\"Count of Mapped Reads (Billions)\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = mapped_reads_perc.mapped_perc.hist(alpha=0.9, ax=ax)\n",
    "ax.set_title(\"Distribution of Mapped Reads (%)\",size=15)\n",
    "ax.set_xlabel(\"Percent Mapped Reads\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_mapped_pct=mapped_reads_perc.loc[:,['mapped_perc','Outlier']]. \\\n",
    "    rename(columns={\"mapped_perc\":\"mapped_pct\",\"Outlier\":\"mapped_pct_outlier\"})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count and Percent of reads properly paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"properly_paired_count\"\n",
    "properly_paired_count = bq.Query('SELECT sample, ' + \n",
    "                             'CAST(value AS FLOAT64) as mapped_count ' \n",
    "                             'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                             'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "properly_paired_perc=qc_passed_reads_count.set_index(['sample']).join(properly_paired_count.set_index(['sample']))\n",
    "properly_paired_perc['mapped_perc']=100*(properly_paired_perc['mapped_count']/properly_paired_perc['total_count'])\n",
    "properly_paired_perc['total_count']=properly_paired_perc.total_count.astype(int)\n",
    "properly_paired_perc['mapped_count']=properly_paired_perc.mapped_count.astype(int)\n",
    "properly_paired_perc['mapped_perc']=properly_paired_perc.mapped_perc.astype(float).round(2)\n",
    "\n",
    "#print('Mean of properly_paired: {0:0.2f}'.format(np.mean(properly_paired_perc.mapped_perc)))\n",
    "#print('Median of properly_paired: {0:0.2f}'.format(np.median(properly_paired_perc.mapped_perc)))\n",
    "#print('STD of properly_paired: {0:0.2f}'.format(np.std(properly_paired_perc.mapped_perc)))\n",
    "#properly_paired_perc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Statistics : Mean, STD, the number of samples with less than 96% properly paired percent "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "std=properly_paired_perc.mapped_perc.std()\n",
    "mean=properly_paired_perc.mapped_perc.mean()\n",
    "print('mean: {0},  std: {1}'.format(round(mean,2), round(std,2)))\n",
    "\n",
    "# if the measurement is 3 times STD from the mean, then flag it\n",
    "# properly_paired_perc['Outlier'] = properly_paired_perc.mapped_perc.apply(lambda x: False if abs(mean-x)<3*std else True)\n",
    "properly_paired_perc['Outlier'] = properly_paired_perc.mapped_perc.apply(lambda x: False if x > 96 else True)\n",
    "#display(properly_paired_perc.head())\n",
    "properly_paired_perc = properly_paired_perc[['mapped_count','mapped_perc','Outlier']]\n",
    "properly_paired_perc.sort_values(by=['mapped_perc'],inplace=True)\n",
    "print('The number of samples in < 96% : {0}'.format(properly_paired_perc[properly_paired_perc.Outlier].shape[0]))\n",
    "print(properly_paired_perc[properly_paired_perc.Outlier])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Distribution of properly paired reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = properly_paired_perc.mapped_count.hist(alpha=0.9, ax=ax)\n",
    "ax.set_title(\"Distribution of Properly Paired Reads (Count)\")\n",
    "ax.set_xlabel(\"Count of Properly Paired Reads (Billions)\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = properly_paired_perc.mapped_perc.hist(alpha=0.9, ax=ax)\n",
    "ax.set_title(\"Distribution of Properly Paired Reads (%)\",size=15)\n",
    "ax.set_xlabel(\"Percent Properly Paired Reads\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_paired_pct=properly_paired_perc.loc[:,['mapped_perc','Outlier']]. \\\n",
    "    rename(columns={\"mapped_perc\":\"paired_pct\",\"Outlier\":\"paired_pct_outlier\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VerifyBamID Results (n=288)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_verifybamid_table = account['GOOGLE_CLOUD_PROJECT']+'.'+account['BIGQUERY_DATASET']+'.check_contamination'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "check_contamination = bq.Query('SELECT SEQ_ID, ' + 'CAST(FREEMIX AS FLOAT64) as contamination_rate ' \n",
    "                             'FROM `' + mvp_verifybamid_table + '` ').execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('The number of samples : ' + str(len(check_contamination)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,6))\n",
    "hist = check_contamination.contamination_rate.plot.box(ax=ax)\n",
    "ax.set_title(\"Box plot of Contamination Rate (%)\",size=15)\n",
    "ax.set_xlabel(\"Contamination Rate\",size=15)\n",
    "ax.set_ylabel(\"Frequency\",size=15)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cut the window in 2 parts\n",
    "f, (ax_box, ax_hist) = plt.subplots(2, sharex=True, gridspec_kw={\"height_ratios\": (.15, .85)}, figsize=(10,6))\n",
    " \n",
    "# Add a graph in each part\n",
    "sns.boxplot(check_contamination[\"contamination_rate\"], ax=ax_box)\n",
    "sns.distplot(check_contamination[\"contamination_rate\"], ax=ax_hist)\n",
    " \n",
    "# Remove x axis name for the boxplot\n",
    "ax_box.set(xlabel='')\n",
    "ax_box.set_title(\"Distribution of Contamination Rate (%)\",size=15)\n",
    "ax_hist.set_xlabel(\"Contamination Rate\",size=15)\n",
    "ax_hist.set_ylabel(\"Frequency\",size=15)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To BigQuery table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_contam=check_contamination.rename(columns={\"SEQ_ID\":\"sample\",\"contamination_rate\":\"contam_rate\"}).set_index('sample')\n",
    "\n",
    "bq_contam.loc[bq_contam.contam_rate < 0.05, 'contam_rate_outlier']=False\n",
    "bq_contam.loc[bq_contam.contam_rate >= 0.05, 'contam_rate_outlier']=True\n",
    "\n",
    "#bq_contam.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BigQuery Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge all QC features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_qc=bq_gc.merge(bq_avg_seq_qual, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_seq_len, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_snps, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_indels, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_ti_tv, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_het_hom, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_mapped_pct, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_paired_pct, how='outer',left_index=True,right_index=True) \\\n",
    "    .merge(bq_contam, how='outer',left_index=True,right_index=True)\n",
    "bq_qc.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add outlier list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## pass\n",
    "bq_qc['outlier']='pass'\n",
    "\n",
    "## fail check\n",
    "bq_qc.loc[(bq_qc['gc_outlier']==True)|(bq_qc['avg_seq_qual_outlier']==True)|(bq_qc['seq_len_outlier']==True) \\\n",
    "          |(bq_qc['snps_outlier']==True)|(bq_qc['indels_outlier']==True)|(bq_qc['ti_tv_outlier']==True) \\\n",
    "          |(bq_qc['het_hom_outlier']==True)|(bq_qc['mapped_pct_outlier']==True)|(bq_qc['paired_pct_outlier']==True) \\\n",
    "          |(bq_qc['contam_rate_outlier']==True),'outlier']='outlier'\n",
    "\n",
    "## outlier\n",
    "bq_qc.loc[(bq_qc['gc_outlier'].isna())|(bq_qc['avg_seq_qual_outlier'].isna())|(bq_qc['seq_len_outlier'].isna()) \\\n",
    "          |(bq_qc['snps_outlier'].isna())|(bq_qc['indels_outlier'].isna())|(bq_qc['ti_tv_outlier'].isna()) \\\n",
    "          |(bq_qc['het_hom_outlier'].isna())|(bq_qc['mapped_pct_outlier'].isna())|(bq_qc['paired_pct_outlier'].isna()) \\\n",
    "          |(bq_qc['contam_rate_outlier'].isna()),'outlier']='uncompleted'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_qc.drop_duplicates(keep='first',inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Outlier and uncompleted lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bq_qc.loc[bq_qc['outlier']=='outlier',:].to_csv('gs://gbsc-gcp-project-mvp-test-trellis/analysis-notebooks/qc_outlier-v058-1.csv')\n",
    "bq_qc.loc[bq_qc['outlier']=='uncompleted',:].to_csv('gs://gbsc-gcp-project-mvp-test-trellis/analysis-notebooks/qc_uncompleted-v058-1.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Upload DF to BigQuery"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_id=account['BIGQUERY_DATASET']+'.qc_analysis'\n",
    "projectid=account['GOOGLE_CLOUD_PROJECT']\n",
    "\n",
    "pandas_gbq.to_gbq(\n",
    "    bq_qc, table_id, project_id=projectid, if_exists='replace',\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
